### PR TITLE
Add pre-commit rules

### LINKED ISSUE #7300: pre-commit rules in tektoncd/pipeline repository
### Feature request

As a tekton contributor it's kind of hard to know which linters and which makefile rules needs to be executed before sending a Pull Request to tektoncd/pipeline

It would be great if we can standardize those rules via [pre-commit](https://pre-commit.com/) which is a popular tools in opensource repositories to make sure before submitting the PR we test it locally first.

We would not need to introduce new pre-commit hooks, just have custom hooks launching the appropriate make commands which reflects what happen in CI (outer loop)
 
Here is an example how we do this on pipelines-as-code:

https://github.com/openshift-pipelines/pipelines-as-code/blob/37ec9a9b7b1da4f6c1eba37cf94ea67d4a37594e/.pre-commit-config.yaml#L36-L70

### Use case

* make it easy for new contributors to know what to run before creating a PR.
* don't have to have a new CI run consuming resources just for a linter errors

### PR DESCRIPTION
**Pull request by roman-kiselenko**
# Changes
This PR introduces the pre-commit hooks in tektoncd/pipeline repository. 
A badge in `README.md`, a note in `DEVELOPMENT.md`, and in `pull_request_template.`

Hooks enabled:

Local:

- [x] make golangci-lint
- [x] make test-unit

Community:

- [x] [check-added-large-files](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_added_large_files.py)
- [x] [check-yaml](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_yaml.py)
- [x] [check-shebang-scripts-are-executable](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_executables_have_shebangs.py)
- [x] [check-json](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_json.py)
- [x] [detect-private-key](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/detect_private_key.py)

Fix #7300

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+repos:
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v4.2.0
+  hooks:
+  - id: check-added-large-files
+  - id: check-yaml
+    args: ["--allow-multiple-documents"]
+    # Ignore yaml files with {{ }} cause they're not valid.
+    exclude: '(test\/gohelloworld\/gohelloworld-chart\/templates\/.*yaml$)'
+  - id: check-shebang-scripts-are-executable
```
Comment (Yongxuanzhang) [inline .pre-commit-config.yaml:14]: I'm curious how this hook know which scripts need to be checked? Does it scan all the files' content?

Comment (roman-kiselenko) [reply .pre-commit-config.yaml:14]: No, AFAIK it perform [checks](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_executables_have_shebangs.py) only against currently staged files
and only if file starts with shebang with respect of [`git config fileMode`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corefileMode)

**Comment by chmouel** [comment]
/lgtm

Looking Good! 🤙🏽

**Comment by roman-kiselenko** [comment]
Request review @tektoncd/maintainers 

Does anything need to be done in order to merge this PR?

**Review by vdemeester — APPROVED** [review]
I'll mark this one as approved. It's not enabled by default, so it's an opt-in for developer — I think we are ok merging this like that 👼🏼

**Pull request merged**